File: 2554-minimum-total-distance-traveled/minimum-total-distance-traveled.py
from functools import lru_cache
from typing import List

class Solution:
    def minimumTotalDistance(self, robot: List[int], factory: List[List[int]]) -> int:
        # Step 1: Sort robots and factories by position
        robot.sort()
        factory.sort()

        # Step 2: Flatten factory positions based on their capacity limits
        # If a factory at pos 2 has limit 3, we treat it as [2, 2, 2]
        factory_positions = []
        for pos, limit in factory:
            factory_positions.extend([pos] * limit)

        m, n = len(robot), len(factory_positions)

        # Step 3: Recursion with memoization
        @lru_cache(None)
        def solve(ri: int, fi: int) -> int:
            # Base Case: All robots are assigned
            if ri == m:
                return 0
            # Base Case: No more factory spots left, but robots remain
            if fi == n:
                return float('inf')

            # Option 1: Assign current robot to the current factory spot
            take = abs(robot[ri] - factory_positions[fi]) + solve(ri + 1, fi + 1)

            # Option 2: Skip this factory spot for the current robot
            skip = solve(ri, fi + 1)

            return min(take, skip)

        result = solve(0, 0)
        
        # Clear cache to prevent memory leaks in environments with multiple test cases
        solve.cache_clear()
        return result

File: 2554-minimum-total-distance-traveled/minimum-total-distance-traveled.js
var minimumTotalDistance = function (robot, factory) {
    const solve = (ri, fi, positions, t) => {
        if (ri >= robot.length) {
            return 0;
        }
        if (fi >= positions.length) {
            return 1e12; // large number to represent impossible case
        }
        if (t[ri][fi] !== -1) {
            return t[ri][fi];
        }

        // Option 1: assign current robot to current factory position
        const take_current_factory = Math.abs(robot[ri] - positions[fi]) + solve(ri + 1, fi + 1, positions, t);

        // Option 2: skip current factory position
        const skip = solve(ri, fi + 1, positions, t);

        // Memoize and return
        return (t[ri][fi] = Math.min(take_current_factory, skip));
    };

    // Step 1: sort robots and factories by position
    robot.sort((a, b) => a - b);
    factory.sort((a, b) => a[0] - b[0]);

    // Step 2: expand factory positions according to their limits
    const positions = [];
    for (let i = 0; i < factory.length; i++) {
        const [pos, limit] = factory[i];
        for (let j = 0; j < limit; j++) {
            positions.push(pos);
        }
    }

    // Step 3: initialize memoization table
    const m = robot.length;
    const n = positions.length;
    const t = Array.from({ length: m + 1 }, () => Array(n + 1).fill(-1));

    // Step 4: recursion with memoization
    return solve(0, 0, positions, t);
};